### DISCUSSION TITLE
providing auth for multiple syncthing instances

### DISCUSSION THREAD

**Discussion by digtail** [kanidm/kanidm#2438] [Q&A]
I am running multiple syncthing instances on multiple nodes. I want to have a single sign on workflow with kanidm as the authentication provider. My current solution is complicated and brittle, as it involves many components.

My setup consists of a central kanidm instance and multiple nodes each running syncthing, a reverse proxy, and oauth2\_proxy.

Syncthing itself does not do oauth2, which is why I'm using oauth2\_proxy to do the oauth2 flow. oauth2\_proxy supports nginx's auth\_mod interface. In my setup, oauth2\_proxy does not act as a proxy itself. Instead the reverse proxy asks oauth2\_proxy whether a request is authrized using the nginx auth\_request protocol.

Ideally kanidm would support auth\_request and multiple domains in oauth2\_rs\_origin.

- If kanidm does auth\_request, I can drop all the oauth2\_proxy instances, which would remove a lot of complexity.
- If kanidm does multiple domains in oauth2\_rs\_origin, I can use one oauth2 proxy in total instead of one per node. This would remove some complexity. I had this setup before, with a gitlab instance as oauth2 provider. (On a personal note, it makes sense to me to represent all the syncthing instances in one kanidm oauth2 configuration, since the syncthing instances all share the same data and I see no point in only giving access to some of them (quite the opposite actually))

Possible alternative solutions include:
* Implement oauth2 in Syncthing.
  * multiple origins in `oauth2_rs_origin` would still be useful for that
  * Enforcing oauth2 at a higher level is a good fallback to have for other applications that don't support oauth2, as implementing oauth2 is often not easy.
* Implement oauth2_proxy functionality in my reverse proxy (feels like the wrong approach to me, would rather implement auth_request in kanidm)

While I currently have a working solution, I would like it to become less complex. I am not sure how that will looks like, as that heavily depends on the features and changes kanidm accepts.

Is there already better solution that I don't see? Is kanidm interested in implementing auth_request? Is this [enough reason](https://github.com/kanidm/kanidm/issues/1290) for kanidm to support multiple domains in oauth2_rs_origin?

**Comment by Firstyear** [comment]
> My current solution is complicated and brittle, as it involves many components.

Yeah, I think sometimes there is an argument for simplicity also meaning you have a reliable service.

> Ideally kanidm would support auth_request and multiple domains in oauth2_rs_origin.

I don't think we want to support auth_request, it appears to be hack to avoid doing things properly. It's also a symptom of the problem (overcomplicated stack, which you yourself noted) rather than a real solution.


As said, the issue here I don't think is Kani, but the way you configured these services. You need to ask yourself "what are you gaining" by configuring your systems like this?

So first, when you have multiple reverse proxies or load balancers, you generally don't ask your users "oh yes, if a.example.com is down just go to b.example.com". HA for reverse proxies is a solved problem with things like anycast or CARP/VRRP. ( https://www.redhat.com/sysadmin/ha-cluster-linux ). Anycast is more work to configure, but is active-active, where as CARP/VRRP is active-passive failover. 

Once you get to this point, you won't need multiple origins, because there is only one origin to your reverse proxy cluster.

Now your application when hosted behind that reverse proxy either:

* the reverse proxy does oauth2 on behalf of the application behind it and passes an authorized header
* is capable of doing it's own authentication, and managing it's own state

So here, either syncthing needs to do oauth2, or your proxy does. You already have oauth2-proxy in the picture, so why not use it as intended ( https://github.com/oauth2-proxy/oauth2-proxy#docs )? Alternately, you could just use Apache2 which does mod_auth_oidc *and* acts as a reverse proxy all in one. 

Additionally this is part of the reason we haven't implemented multiple origin support - every request for it turns out to be a infrastructure design/configuration issue and multiple-origins is trying to "patch over" that. Where as once you consider that you only have a single origin, and you configure your infrastructure in a simpler and more durable way, you end up with a better system.

The other reason is practical - from the user application portal we only have a *single url we can redirect the user to*. If you have multiple origins, how are we meant to know which to redirect the user too? It's not Kani's job to load-balance and monitor your systems to know which origin to redirect you to. 

If you have multiple separate and distinct applications behind your reverse proxy that's *more* reason not to support multiple urls - you need them all as seperate oauth2rs configs in kanidm so that each application appeals in the app portal. If you configured syncthing and say nextcloud both as the same oauth2 client, we'll only show one in the app portal. 

So I'd advise you re-architect your proxy setup here. You have all the tools already in place to do this in a simpler and more durable way, without us resorting to hacks like auth_request.

**Comment by yaleman** [comment]
Just to understand the current design - what are the individual instances of syncthing for? One instance per server you're backing up, or for HA?

**Comment by digtail** [comment]
Thank you for your responses.

This is not a high availability setup. Accessing the syncthing instances individually is important.

I  will try to implement oauth2 in my reverse proxy or consider a replacement (once I find the time).
